### PR TITLE
Schema migration guide

### PR DESCRIPTION
**Pull request by rbans96**
## Description
Adds the documentation for Schema migration.

## Related issues
Addresses #74591

## Testing
NR

### CODE REVIEW

File: docs/SchemaMigrationGuide.md [hunk 1]
```diff
@@ -1,98 +1,89 @@
 # Schema Migration Guide
-This guide describes how to upgrade the schema for the FHIR server. The upgrade scripts are the T-SQL script that alters the database in some way and has a version encoded in its filename. The version will be an incrementing integer.
+This document describes how to perform Sql schema migration on the Open source FHIR server.
```
Comment (LTA-Thinking) [inline docs/SchemaMigrationGuide.md:2]: In the readme we use FHIR Server for Azure to refer to the project. Should we use that here too?

File: docs/SchemaMigrationGuide.md [hunk 2]
```diff
@@ -1,98 +1,89 @@
...
 
-#### -- version
+    - #### Test, test
 
-It applies all the versions between current version and the specified version.
+        Test before executing the migration and test after the migration is completed.
 
-#### --latest
+ ## How to toggle between the options
+There is a configurable property in the launchSettings.json which can be set to true/false based on the desired behavior.
```
Comment (LTA-Thinking) [inline docs/SchemaMigrationGuide.md:65]: What is the default value?

File: docs/SchemaMigrationGuide.md [hunk 3]
```diff
@@ -1,98 +1,89 @@
...
-|current|Returns the current versions from the SchemaVersion table along with information on the instances using the given version|--server|
-|available|Returns the versions greater than or equal to the current version along with links to the T-SQL scripts for upgrades|--server|
-|apply|Applies the specified version(s) to the connection string supplied. Optionally can poll the FHIR server current version to apply multiple versions in sequence|--server,<br /> --connection-string,<br /> --next,<br /> --version,<br /> --latest,<br /> --force|
+- ### Schema migration tool
 
-#### Options 
+    The detailed information about this tool is provided [here](SchemaMigrationTool.md).
 
-#### --next
+ - ### Best practices for schema upgrade on Sql server
```
Comment (LTA-Thinking) [inline docs/SchemaMigrationGuide.md:50]: Should this be its own top level section? I think the advice in it is relevant to both the automatic and manual upgrade.

File: docs/SchemaMigrationTool.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
+# Schema Migration Tool
+The Schema migration tool is the command line utility to perform schema upgrade on demand. The tool will automatically identify current version of the database in order to apply appropriate migration scripts.
```
Comment (LTA-Thinking) [inline docs/SchemaMigrationTool.md:2]: Nit: missing 'the' between 'identify' and 'current version'

File: docs/SchemaMigrationTool.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    The tool would only run when the current schema is in conformance to the [BaseSchema](BaseSchema.md). So either the BaseSchema should already be present or if not, then Schema admin would need to run BaseSchema script manually.
+
+- #### How to install the tool
+
+    - ##### If available as .NET Core global tool 
+
+        It can be installed like this:
+
+            Open a terminal/command prompt 
+            Type 'dotnet tool install -g Microsoft.Health.SchemaManager'
```
Comment (LTA-Thinking) [inline docs/SchemaMigrationTool.md:16]: This section is being displayed like it was code. I don't think this is the right style for the section.

File: docs/SchemaMigrationTool.md [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+        It can be installed like this:
+            
+            Visual Studio setup - On the Tools menu, select Options > NuGet Package Manager > Package Sources. Select the green plus in the upper-right corner and enter the name and source URL as below:
+
+            Name: InternalBuilds
+            Source: https://microsofthealthoss.pkgs.visualstudio.com/_packaging/InternalBuilds/nuget/v3/index.json
+            
+            In the Package Manager Console, type the below command
+            PM> dotnet tool install -g Microsoft.Health.SchemaManager --add-source ./ --version [latestversion]
```
Comment (LTA-Thinking) [inline docs/SchemaMigrationTool.md:28]: Same comment as above. Parts of the section should display as code, but not the whole thing.

**Comment by LTA-Thinking** [comment]
The docs mention that if a server is multiple versions behind it can upgrade by sequentially updating the server through each version. What happens if the server fails one of the upgrades after successfully completing at least one? Are the successful ones rolled back, or is the server left at the last version it successfully upgraded to?

File: docs/BaseSchema.md [hunk 7]
```diff
@@ -0,0 +1,174 @@
+# BaseSchema
+
+The following are the required schema for Schema migration tool. It includes the couple of tables and stored procedures corresponding to these tables.
```
Comment (LTA-Thinking) [inline docs/BaseSchema.md:3]: Could you make 'Schema migration' a hyperlink to SchemaMigrationTool.md?

Comment (LTA-Thinking) [inline docs/BaseSchema.md:3]: Nit: I don't think you need 'couple of' before 'tables and stored procedures'.

File: docs/BaseSchema.md [hunk 8]
```diff
@@ -0,0 +1,174 @@
+# BaseSchema
```
Comment (LTA-Thinking) [inline docs/BaseSchema.md:1]: Are these tables and stored procedures something a DBA would have to make manually the first time they upgrade a server, or do we have a tool to make them?

File: docs/SchemaMigrationGuide.md (hunk 2 above)
Comment (rbans96) [reply docs/SchemaMigrationGuide.md]: The default is false. Updated the doc

**Comment by rbans96** [comment]
> The docs mention that if a server is multiple versions behind it can upgrade by sequentially updating the server through each version. What happens if the server fails one of the upgrades after successfully completing at least one? Are the successful ones rolled back, or is the server left at the last version it successfully upgraded to?

Good question. If the specified version is compatible and apply execution fails after successfully completing at least one, then the successfully executed version won't rollback.

File: docs/BaseSchema.md (hunk 8 above)
Comment (rbans96) [reply docs/BaseSchema.md:1]: Good question! As mentioned in the doc, for now DBA would need to execute it manually if it doesn't exist https://github.com/microsoft/fhir-server/pull/1120/files#diff-88920ff7161e7d92d1ae16e8c15c646fR7

But we had thought about it and might create a backlog item to execute the BaseSchema by the tool itself.

**Pull request merged**